src/redprl/refiner.fun

```diff
@@ -1015,8 +1015,224 @@ struct
       end
   end
 
+  structure Restriction =
+  struct
+    (* This structure provides functions that automate the restriction
+       judgement rules given in "Dependent Cubical Realizability",
+       page 46.
+
+       Such automation is possible because the rules are "syntax
+       directed" on the restriction being performed.
+     *)
+
+    (* Return whether given dimensions are different constants, which
+       is a contradiction, allowing the restriction judgment to conclude
+       anything. *)
+    fun dimsContradictory (P.APP P.DIM0, P.APP DIM1) = true
+      | dimsContradictory (P.APP P.DIM1, P.APP DIM0) = true
+      | dimsContradictory _ = false
+
+    (* Restrict a judgement by a single equation `ext = eps`. *)
+    fun One (jdg : abt jdg) (ext : param) (eps : param) =
+      if P.eq Sym.eq (ext, eps) then [jdg] (* combining rules in first row *)
+      else if dimsContradictory (ext, eps) then [] (* second row, left rule *)
+      else
+        let
+          val (P.VAR v, P.APP eps) = (ext, eps)
+        in
+          [Seq.map (substSymbol (P.APP eps, v)) jdg] (* bottom left rule *)
+        end
+
+    (* Restrict a judgement by two equations `ext0 = eps0` and `ext1 = eps1`. *)
+    fun Two (jdg : abt jdg) (ext0 : param) (eps0 : param) (ext1 : param) (eps1 : param) =
+      if P.eq Sym.eq (ext0, eps0) then One jdg ext1 eps1 (* top right rule *)
+      else if P.eq Sym.eq (ext1, eps1) then One jdg ext0 eps0 (* top right rule *)
+      else if dimsContradictory (ext0, eps0) then [] (* second row, left rule *)
+      else if dimsContradictory (ext1, eps1) then [] (* second row, left rule *)
+      else
+        let
+          val (P.VAR u, P.APP _) = (ext0, eps0)
+          val (P.VAR v, P.APP _) = (ext1, eps1)
+        in
+          if Sym.eq (u, v) andalso not (P.eq Sym.eq (eps0, eps1)) then [] (* second row, right rule *)
+          else
+            [Seq.map (substSymbol (eps0, u)) (Seq.map (substSymbol (eps1, v)) jdg)] (* bottom right rule *)
+        end
+  end
+
+  structure HCom =
+  struct
+    (* Given a list of n extents and a list of 2n tubes, produce a list of
+       (extent * side * tube) triples, where side is 0 or 1 according to the
+       tube's position in the list *)
+    fun groupTubes [] [] = []
+      | groupTubes (ext :: exts) (tube0 :: tube1 :: tubes)  =
+        (ext, P.APP P.DIM0, tube0) :: (ext, P.APP P.DIM1, tube1) :: groupTubes exts tubes
+      | groupTubes _ _ = raise Fail "groupTubes"
+
+    fun listToTel l = List.foldl (fn (g, l) => l >: g) T.empty l
+
+    (* Produce the list of goals requiring that tube aspects agree with each other. *)
+    fun intraTubeGoals H w ty group =
```

Comment:
Yes, that's right. We do need to consider all pairs of tube faces. The local function `intraTube` produces the (one or zero) goals corresponding to a particular choice of tube faces. Then we wrap that up in a couple `ListMonad.bind`s to get all of them. 

Does that seem right?